### PR TITLE
context_output for  vector store retriever 

### PR DESCRIPTION
**Pull request by Hansrajr**
## Proposed changes

This PR include changes for the followings:

1. it  updates events for context_output for langchain, llama_index and haystack retriever span.
2. get the sagemaker endpoint for aws lambda and sets it to vector store retriever span attributes as embedding_model.

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/monocle2ai/monocle/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/monocle_apptrace/wrap_common.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import logging
 import os
 from urllib.parse import urlparse
-
+from langchain_community.embeddings import SagemakerEndpointEmbeddings
```
[the remark this answers is not in the document]
Comment (kshitiz-okahu) [reply src/monocle_apptrace/wrap_common.py:5]: @Hansrajr

[the remark this answers is not in the document]
Comment (Hansrajr) [reply src/monocle_apptrace/wrap_common.py]: done @kshitiz-okahu

File: src/monocle_apptrace/wrap_common.py [hunk 2]
```diff
@@ -29,15 +29,40 @@
...
 }
 
+
+def get_embedding_model_for_vectorstore(instance):
+    # Handle Langchain or other frameworks where vectorstore exists
+    if hasattr(instance, 'vectorstore'):
+        vectorstore_dict = instance.vectorstore.__dict__
+
+        # Check if SagemakerEndpointEmbeddings is present
+        if 'embedding_func' in vectorstore_dict and isinstance(vectorstore_dict['embedding_func'], SagemakerEndpointEmbeddings):
```
[the remark this answers is not in the document]
Comment (Hansrajr) [reply src/monocle_apptrace/wrap_common.py]: Done @kshitiz-okahu

[the remark this answers is not in the document]
Comment (Hansrajr) [reply src/monocle_apptrace/wrap_common.py]: output from okahu demo sagmaker app using monocle_apptrace 

{
    "name": "langchain.task.VectorStoreRetriever",
    "context": {
        "trace_id": "0x18a697d825e7435659048365e55f1488",
        "span_id": "0x932d97907b4421c7",
        "trace_state": "[]"
    },
    "kind": "SpanKind.INTERNAL",
    "parent_id": "0x7c347864f86a5994",
    "start_time": "2024-10-03T14:07:08.578604Z",
    "end_time": "2024-10-03T14:07:10.234005Z",
    "status": {
        "status_code": "UNSET"
    },
    "attributes": {
        "tags": [
            "Milvus",
            "SagemakerEndpointEmbeddings"
        ],
        **"type": "vector_store",
        "provider_name": "Milvus",
        "embedding_model": "okahu-sagemaker-rag-embedding-ep"
    },**
    **"events": [
        {
            "name": "context_input",
            "timestamp": "2024-10-03T14:07:08.579181Z",
            "attributes": {
                "question": "Which name is also used to describe the Amazon rainforest in English?"
            }
        },
        {
            "name": "context_output",
            "timestamp": "2024-10-03T14:07:10.233964Z",
            "attributes": {
                "response": "The Amazon rainforest (Portuguese: Floresta Amazonica or Amazonia; Spanish: Selva Amazonica, Amazon\u00ed..."
            }
        }**
    ],
    "links": [],
    "resource": {
        "attributes": {
            "service.name": "sagemaker_workflow_1"
        },
        "schema_url": ""
    }
}

File: src/monocle_apptrace/utils.py [hunk 3]
```diff
@@ -91,3 +92,12 @@ def get_embedding_model() -> str:
     @return: The name of the embedding model, or 'unknown' if not set
     """
     return embedding_model_context.get('embedding_model', 'unknown')
+
+
+def set_context_input(inputs: str, input_key: str):
+    # Set the workflow input in the global context
+    attach(set_value(input_key, inputs))
+
+def get_context_input(input_key: str) -> str:
```
Comment (oi-raanne) [inline src/monocle_apptrace/utils.py:101]: setter and getter should be for all attributes.

**Review by oi-raanne — CHANGES_REQUESTED** [review]

File: src/monocle_apptrace/utils.py (hunk 3 above)
Comment (Hansrajr) [reply src/monocle_apptrace/utils.py]: done

File: src/monocle_apptrace/wrap_common.py [hunk 4]
```diff
@@ -257,21 +291,36 @@ def update_workflow_type(to_wrap, span: Span):
         if(package_name is not None and package in package_name):
             span.set_attribute(WORKFLOW_TYPE_KEY, workflow_type)
 
-def update_span_with_context_input(to_wrap, wrapped_args ,span: Span):
+def update_span_with_context_input(to_wrap, wrapped_args, instance ,span: Span):
     package_name: str = to_wrap.get('package')
     if "langchain_core.retrievers" in package_name:
         input_arg_text = wrapped_args[0]
         span.add_event(CONTEXT_INPUT_KEY, {QUERY:input_arg_text})
```
Comment (kshitiz-okahu) [inline src/monocle_apptrace/wrap_common.py]: Can you remove the duplication of the following line
span.add_event(CONTEXT_INPUT_KEY, {QUERY:input_arg_text})

File: src/monocle_apptrace/wrap_common.py [hunk 5]
```diff
@@ -257,21 +291,36 @@ def update_workflow_type(to_wrap, span: Span):
...
+        span.add_event(CONTEXT_INPUT_KEY, {QUERY:input_arg_text})
 
 def update_span_with_context_output(to_wrap, return_value ,span: Span):
     package_name: str = to_wrap.get('package')
+    if "langchain_core.retrievers" in package_name:
+        combined_output = " ".join([doc.page_content for doc in return_value if hasattr(doc, 'page_content')])
+        if len(combined_output) > 100:
+            combined_output = combined_output[:100] + "..."
+
+        span.add_event(CONTEXT_OUTPUT_KEY, {RESPONSE: combined_output})
```
Comment (kshitiz-okahu) [inline src/monocle_apptrace/wrap_common.py:313]: Can you remove duplication of 
span.add_event(CONTEXT_OUTPUT_KEY, {RESPONSE: combined_output})

File: src/monocle_apptrace/wrap_common.py (hunk 4 above)
Comment (Hansrajr) [reply src/monocle_apptrace/wrap_common.py]: done

File: src/monocle_apptrace/wrap_common.py (hunk 5 above)
Comment (Hansrajr) [reply src/monocle_apptrace/wrap_common.py]: done

**Review by kshitiz-okahu — APPROVED** [review]

**Review by oi-raanne — APPROVED** [review]

**Pull request merged**